coverage: Move most per-function coverage info into `mir::Body`

Zalathar opened the pull request:
Currently, all of the coverage information collected by the `InstrumentCoverage` pass is smuggled through MIR in the form of individual `StatementKind::Coverage` statements, which must then be reassembled by coverage codegen.

That's awkward for a number of reasons:
- While some of the coverage statements do care about their specific position in the MIR control-flow graph, many of them don't, and are just tacked onto the function's first BB as metadata carriers.
- MIR inlining can result in coverage statements being duplicated, so coverage codegen has to jump through hoops to avoid emitting duplicate mappings.
- MIR optimizations that would delete coverage statements need to carefully copy them into the function's first BB so as not to omit them from coverage reports.
- The order in which coverage codegen sees coverage statements is dependent on MIR optimizations/inlining, which can cause unnecessary churn in the emitted coverage mappings.
- We don't have a good way to annotate MIR-level functions with extra coverage info that doesn't belong in a statement.

---

This PR therefore takes most of the per-function coverage info and stores it in a field in `mir::Body` as `Option<Box<FunctionCoverageInfo>>`.

(This adds one pointer to the size of `mir::Body`, even when coverage is not enabled.)

Coverage statements still need to be injected into MIR in some cases, but only when they actually affect codegen (counters) or are needed to detect code that has been optimized away as unreachable (counters/expressions).

---

By the end of this PR, the information stored in `FunctionCoverageInfo` is:

- A hash of the function's source code (needed by LLVM's coverage map format)
- The number of coverage counters added by coverage instrumentation
- A table of coverage expressions, associating each expression ID with its operator (add or subtract) and its two operands
- The list of mappings, associating each covered code region with a counter/expression/zero value

---

~~This is built on top of #115301, so I'll rebase and roll a reviewer once that lands.~~
r? @ghost
@rustbot label +A-code-coverage

Zalathar committed:
coverage: Rename `Operand` to `CovTerm`

Later patches in this PR will use `CovTerm` to represent things that are not
expression operands.

Zalathar committed:
coverage: Collect a function's coverage mappings into a single list

This is an intermediate step towards being able to store all of a function's
mappings in function coverage info.

Zalathar committed:
Remove unused `TyCtxt` from `remove_dead_blocks`

This context was only needed by code for processing coverage statements, which
has been removed.

Zalathar committed:
coverage: Attach an optional `FunctionCoverageInfo` to `mir::Body`

This allows coverage information to be attached to the function as a whole when
appropriate, instead of being smuggled through coverage statements in the
function's basic blocks.

As an example, this patch moves the `function_source_hash` value out of
individual `CoverageKind::Counter` statements and into the per-function info.

When synthesizing unused functions for coverage purposes, the absence of this
info is taken to indicate that a function was not eligible for coverage and
should not be synthesized.

Zalathar committed:
coverage: Store the number of counters/expressions in function coverage info

Coverage codegen can now allocate arrays based on the number of
counters/expressions originally used by the instrumentor.

The existing query that inspects coverage statements is still used for
determining the number of counters passed to `llvm.instrprof.increment`. If
some high-numbered counters were removed by MIR optimizations, the instrumented
binary can potentially use less memory and disk space at runtime.

Zalathar committed:
coverage: Store all of a function's mappings in function coverage info

Previously, mappings were attached to individual coverage statements in MIR.
That necessitated special handling in MIR optimizations to avoid deleting those
statements, since otherwise codegen would be unable to reassemble the original
list of mappings.

With this change, a function's list of mappings is now attached to its MIR
body, and survives intact even if individual statements are deleted by
optimizations.

Zalathar committed:
coverage: Store expression data in function coverage info

Even though expression details are now stored in the info structure, we still
need to inject `ExpressionUsed` statements into MIR, because if one is missing
during codegen then we know that it was optimized out and we can remap all of
its associated code regions to zero.

Zalathar committed:
coverage: Make expression simplification non-destructive

Instead of modifying the accumulated expressions in-place, we now build a set
of expressions that are known to be zero, and then consult that set on the fly
when converting the expression data for FFI.

This will be necessary when moving mappings and expression data into function
coverage info, which can't be mutated during codegen.

Zalathar commented:
When I moved expression data into `FunctionCoverageInfo`, I was hoping to be able to get rid of `CoverageKind::Expression`, leaving counter-increments as the only kind of coverage statement.

To my surprise, I found that this can't be done without regressing the mappings: even though an expression statement no longer carries op/operand data, it still witnesses the fact that its enclosing BB/BCB wasn't removed by later MIR transforms.

(`CoverageKind::Counter` also witnesses non-removal in the same way, in addition to lowering to `llvm.instrprof.increment`.)

So when coverage codegen generates the final mappings to embed in the binary, it can zero out any mappings whose BB/BCB was removed by MIR transforms, while leaving all other mappings intact.

(This zeroing is also currently load-bearing for *unused* functions: instead of explicitly zeroing out all of the mappings, we rely on the normal-case zeroing code to observe that no counters/expressions were ever seen for an unused function's mappings.)

Zalathar commented:
I've also renamed the remaining variants of `CoverageKind` [(diff)](https://github.com/rust-lang/rust/compare/4b27f9de38064c3e6d4ae6c1fc97b7232bc96f50..d990af0eaf917c5b3370212b33d39aa02dfe1ae5), since this PR alters/narrows their roles.

Zalathar committed:
coverage: Update docs for `StatementKind::Coverage`

This new description reflects the changes made in this PR, and should hopefully
be more useful to non-coverage developers who need to care about coverage
statements.

bors commented:
:umbrella: The latest upstream changes (presumably #116154) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Zalathar commented:
I realised that with the debug code now gone (#115962), there's no compelling reason for `CoverageCounters::expressions` to be public, so I made it private and added the necessary accessor methods [(diff)](https://github.com/rust-lang/rust/compare/1d927bf3da46b3024660193daca171a015858575..f122ca449b69cc9739cdb336d9fa5d0f230cca28).

Zalathar commented:
Tweaked the docs for `CoverageKind` variants [(diff)](https://github.com/rust-lang/rust/compare/403c116a9348bfef46c97c9597ef34286a1d230d..0a45515a1fff4585fc1eae0abdc99a409f7b3a27).

Zalathar commented:
#115301 has landed, so this is now ready for review.

@rustbot ready
r? compiler

Zalathar commented:
As usual, I'll keep rebasing this for the benefit of my other work that builds on top of it.

Zalathar commented on compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs:
```diff
@@ -124,11 +124,22 @@ impl<'tcx> CoverageInfoBuilderMethods<'tcx> for Builder<'_, '_, 'tcx> {
                 let fn_name = bx.get_pgo_func_name_var(instance);
                 let hash = bx.const_u64(function_coverage_info.function_source_hash);
-                let num_counters = bx.const_u32(coverageinfo.num_counters);
+                let num_counters = bx.const_u32(num_counters);
```
I had originally wanted to do that, but eventually decided against it, for two reasons:

- LLVM uses this number to determine how many physical counter variables to create in memory and in `.profraw` files. So in cases where high-numbered counters are removed by MIR opts, this reduces the overall size of the profile data. Removing the query now would make profile data larger than before, in cases where it's currently making a difference.
- In the future, I want to improve this to get rid of *all* of the unused counters (instead of just the high-numbered ones), by renumbering the used counters to get rid of holes. If I remove the coverage query plumbing now, I'll just have to add it all back again later.

Zalathar commented on compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs:
```diff
@@ -124,11 +124,22 @@ impl<'tcx> CoverageInfoBuilderMethods<'tcx> for Builder<'_, '_, 'tcx> {
+                // if some high-numbered counters were removed by MIR optimizations.
+                let num_counters =
+                    bx.tcx().coverage_ids_info(instance.def).max_counter_id.as_u32() + 1;
+                if num_counters as usize > function_coverage_info.num_counters {
+                    // But it should never be larger.
+                    bug!(
+                        "num_counters disagreement: query says {num_counters} but function info only has {}",
+                        function_coverage_info.num_counters
+                    );
+                }
```
For some reason I had it in my head that `bug!` produces nicer output. But even if that's true, an `assert!` is probably more appropriate here.

Zalathar commented on tests/mir-opt/instrument_coverage.bar.InstrumentCoverage.diff:
```diff
@@ -5,7 +5,7 @@
       let mut _0: bool;
   
       bb0: {
-+         Coverage::Counter(0) for [/the/src/instrument_coverage.rs:20:1 - 22:2];
```
You're probably right. I hadn't bothered because I mostly rely on `coverage-map` tests and haven't needed that information in MIR dumps (yet), but it will probably be a lifesaver someday.

Zalathar commented on tests/mir-opt/instrument_coverage.main.InstrumentCoverage.diff:
```diff
@@ -13,7 +13,7 @@
   
       bb1: {
-+         Coverage::Expression(0) = Counter(0) + Counter(1);
++         Coverage::ExpressionUsed(0);
```
Yeah, if I add in support for dumping any part of `FunctionCoverageInfo` then I should probably dump the whole thing.

Zalathar commented on compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs:
```diff
@@ -52,16 +46,27 @@ impl<'tcx> FunctionCoverage<'tcx> {
+        let mut expressions_not_seen = BitSet::new_empty(num_expressions);
+        for Mapping { term, .. } in &function_coverage_info.mappings {
+            if let &CovTerm::Expression(id) = term {
+                expressions_not_seen.insert(id);
```
When this set eventually gets used (in `identify_zero_expressions`), we only want it to contain expression IDs that we *expected* to see in MIR (as `ExpressionUsed`), but didn't actually see.

So in this initialization step, we identify the subset of expression IDs that we expect to see (i.e. those that are directly used by one or more regions). During MIR traversal, every `ExpressionUsed` causes its ID to be removed from the set.

Expression IDs that are only used indirectly (by other expressions) don't have a corresponding `ExpressionUsed` statement. If we only tracked “seen” expressions, those indirect expressions would be treated as “unseen”, which isn't what we want, so we have to exempt them somewhere.

So by using upside-down logic for the set, the code that actually uses the set gets to be simpler.

(I'll try to find ways to make this more obvious.)

Zalathar replied:
I ended up deciding to reverse this and move to `expressions_seen`, because that reduced the number of double-negatives I have to use in explaining how any of this works.

Zalathar commented on compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs:
```diff
@@ -145,32 +143,37 @@ impl<'tcx> FunctionCoverage<'tcx> {
+            // We don't need to simplify the actual expression data in the
+            // expressions list; we can just simplify a temporary copy and then
+            // use that to update the set of always-zero expressions.
+            let mut expression = expression.clone();
```
The main reason I used clone here was to avoid having to touch the rest of the code below that uses field-access and struct pattern-matching.

I'll try out destructuring to see whether that looks nicer.

Zalathar commented:
I'll push a rebase first, and then look into making changes based on feedback.

Zalathar commented on compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs:
```diff
@@ -12,28 +13,21 @@ pub struct Expression {
-    counters: IndexVec<CounterId, Option<Vec<CodeRegion>>>,
+
+    /// Tracks which counters have been seen, to avoid duplicate mappings
+    /// that might be introduced by MIR inlining.
```
Before inlining, counter IDs (and expression IDs) are local to their enclosing function.

After inlining, we use source scopes to identify which function a coverage statement came from. This is handled in <https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_ssa/src/mir/coverageinfo.rs>.

So when `FunctionCoverage` sees a coverage statement, it doesn't have to worry about inlining. All of the statements it sees either weren't inlined (and thus belong to its `instance`), or were inlined *from* the function that eventually because its `instance`. So it can assume that the counter/expression IDs it sees are ones that make sense for this instance.

Zalathar replied:
(The fact that counter/expression IDs are function-local might not be explicitly documented anywhere; it's one of those things that became so obvious to me that I didn't think to write it down.)

Zalathar replied:
And yeah, if a `CounterIncrement` or `ExpressionUsed` is inlined into another function, the original instance may see multiple copies of its own counter/expression IDs. (The function being inlined into won’t see those IDs.)

Zalathar committed:
coverage: Explicitly note that counter/expression IDs are function-local

Zalathar commented on compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs:
```diff
@@ -145,32 +143,37 @@ impl<'tcx> FunctionCoverage<'tcx> {
         // (By construction, expressions can only refer to other expressions
-        // that have lower IDs, so one simplification pass is sufficient.)
-        for (id, maybe_expression) in self.expressions.iter_enumerated_mut() {
+        // that have lower IDs, so one pass is sufficient.)
```
This isn't a load-bearing invariant, so if it's wrong we just get bloatier-but-equivalent coverage mappings.

Nevertheless, I'll see about adding an assertion anyway; it should be pretty easy to just check directly in one of these expression loops.

Zalathar commented:
Addressed some feedback [(diff)](https://github.com/rust-lang/rust/compare/b9fd59bb096162fcae90d0b36538f000a9015959..75ebfb6feb87c8217db1cea32b9c919509d5949b).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6557212947/job/17808420560) [(plain)](https://github.com/rust-lang/rust/commit/b064783064246b58b5e7369e136185b06bca834a/checks/17808420560/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=Zalathar
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_d2d9d5b1-6213-4c3e-a16a-74d6e5c9c62d
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=6a0755d87832a9c76eaa0ddc1c060e71e4f18500
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_d2d9d5b1-6213-4c3e-a16a-74d6e5c9c62d
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_d2d9d5b1-6213-4c3e-a16a-74d6e5c9c62d
GITHUB_TRIGGERING_ACTOR=Zalathar
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116046/merge
GITHUB_WORKFLOW_SHA=6a0755d87832a9c76eaa0ddc1c060e71e4f18500
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
---

---- [coverage-map] tests/coverage-map/status-quo/try_error_result.rs stdout ----
diff of cov-map:

1 Function name: <try_error_result::Thing1>::get_thing_2
- Raw bytes (63): 0x[01, 01, 02, 01, 05, 05, 02, 0b, 01, 28, 05, 01, 18, 05, 02, 0d, 00, 14, 00, 00, 0d, 00, 14, 00, 00, 0d, 00, 14, 00, 00, 0d, 00, 14, 00, 00, 0d, 00, 14, 00, 00, 0d, 00, 14, 02, 02, 0d, 00, 1a, 00, 00, 0d, 00, 1a, 00, 00, 0d, 00, 1a, 07, 02, 05, 00, 06]
+ Raw bytes (28): 0x[01, 01, 02, 01, 05, 05, 02, 04, 01, 28, 05, 01, 18, 05, 02, 0d, 00, 14, 02, 02, 0d, 00, 1a, 07, 02, 05, 00, 06]
4 - file 0 => global file 1
5 Number of expressions: 2

6 - expression 0 operands: lhs = Counter(0), rhs = Counter(1)
6 - expression 0 operands: lhs = Counter(0), rhs = Counter(1)
7 - expression 1 operands: lhs = Counter(1), rhs = Expression(0, Sub)
- Number of file 0 mappings: 11
+ Number of file 0 mappings: 4
+ Number of file 0 mappings: 4
9 - Code(Counter(0)) at (prev + 40, 5) to (start + 1, 24)
10 - Code(Counter(1)) at (prev + 2, 13) to (start + 0, 20)
- - Code(Zero) at (prev + 0, 13) to (start + 0, 20)
- - Code(Zero) at (prev + 0, 13) to (start + 0, 20)
- - Code(Zero) at (prev + 0, 13) to (start + 0, 20)
- - Code(Zero) at (prev + 0, 13) to (start + 0, 20)
- - Code(Zero) at (prev + 0, 13) to (start + 0, 20)
16 - Code(Expression(0, Sub)) at (prev + 2, 13) to (start + 0, 26)
17     = (c0 - c1)
- - Code(Zero) at (prev + 0, 13) to (start + 0, 26)
- - Code(Zero) at (prev + 0, 13) to (start + 0, 26)
20 - Code(Expression(1, Add)) at (prev + 2, 5) to (start + 0, 6)
21     = (c1 + (c0 - c1))


23 Function name: <try_error_result::Thing2>::call
- Raw bytes (63): 0x[01, 01, 02, 01, 05, 05, 02, 0b, 01, 33, 05, 01, 18, 05, 02, 0d, 00, 14, 00, 00, 0d, 00, 14, 00, 00, 0d, 00, 14, 00, 00, 0d, 00, 14, 00, 00, 0d, 00, 14, 02, 02, 0d, 00, 13, 00, 00, 0d, 00, 13, 00, 00, 0d, 00, 13, 00, 00, 0d, 00, 13, 07, 02, 05, 00, 06]
+ Raw bytes (28): 0x[01, 01, 02, 01, 05, 05, 02, 04, 01, 33, 05, 01, 18, 05, 02, 0d, 00, 14, 02, 02, 0d, 00, 13, 07, 02, 05, 00, 06]
26 - file 0 => global file 1
27 Number of expressions: 2

28 - expression 0 operands: lhs = Counter(0), rhs = Counter(1)
28 - expression 0 operands: lhs = Counter(0), rhs = Counter(1)
29 - expression 1 operands: lhs = Counter(1), rhs = Expression(0, Sub)
- Number of file 0 mappings: 11
+ Number of file 0 mappings: 4
31 - Code(Counter(0)) at (prev + 51, 5) to (start + 1, 24)
32 - Code(Counter(1)) at (prev + 2, 13) to (start + 0, 20)
- - Code(Zero) at (prev + 0, 13) to (start + 0, 20)
- - Code(Zero) at (prev + 0, 13) to (start + 0, 20)
- - Code(Zero) at (prev + 0, 13) to (start + 0, 20)
- - Code(Zero) at (prev + 0, 13) to (start + 0, 20)
37 - Code(Expression(0, Sub)) at (prev + 2, 13) to (start + 0, 19)
38     = (c0 - c1)
- - Code(Zero) at (prev + 0, 13) to (start + 0, 19)
- - Code(Zero) at (prev + 0, 13) to (start + 0, 19)
- - Code(Zero) at (prev + 0, 13) to (start + 0, 19)
42 - Code(Expression(1, Add)) at (prev + 2, 5) to (start + 0, 6)
43     = (c1 + (c0 - c1))

45 Function name: try_error_result::call
45 Function name: try_error_result::call
- Raw bytes (43): 0x[01, 01, 02, 01, 05, 05, 02, 07, 01, 04, 01, 01, 14, 05, 02, 09, 00, 10, 00, 00, 09, 00, 10, 00, 00, 09, 00, 10, 02, 02, 09, 00, 0f, 00, 00, 09, 00, 0f, 07, 02, 01, 00, 02]
+ Raw bytes (28): 0x[01, 01, 02, 01, 05, 05, 02, 04, 01, 04, 01, 01, 14, 05, 02, 09, 00, 10, 02, 02, 09, 00, 0f, 07, 02, 01, 00, 02]
48 - file 0 => global file 1
49 Number of expressions: 2

50 - expression 0 operands: lhs = Counter(0), rhs = Counter(1)
50 - expression 0 operands: lhs = Counter(0), rhs = Counter(1)
51 - expression 1 operands: lhs = Counter(1), rhs = Expression(0, Sub)
- Number of file 0 mappings: 7
+ Number of file 0 mappings: 4
53 - Code(Counter(0)) at (prev + 4, 1) to (start + 1, 20)
54 - Code(Counter(1)) at (prev + 2, 9) to (start + 0, 16)
- - Code(Zero) at (prev + 0, 9) to (start + 0, 16)
- - Code(Zero) at (prev + 0, 9) to (start + 0, 16)
57 - Code(Expression(0, Sub)) at (prev + 2, 9) to (start + 0, 15)
58     = (c0 - c1)
- - Code(Zero) at (prev + 0, 9) to (start + 0, 15)
60 - Code(Expression(1, Add)) at (prev + 2, 1) to (start + 0, 2)
61     = (c1 + (c0 - c1))


The actual cov-map differed from the expected cov-map.
The actual cov-map differed from the expected cov-map.
Actual cov-map saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/coverage-map/status-quo/try_error_result/try_error_result.cov-map
error: 1 errors occurred comparing coverage output.
status: exit status: 0
status: exit status: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/coverage-dump" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/coverage-map/status-quo/try_error_result/try_error_result.ll"
--- stdout -------------------------------
Function name: <try_error_result::Thing1>::get_thing_2
Raw bytes (28): 0x[01, 01, 02, 01, 05, 05, 02, 04, 01, 28, 05, 01, 18, 05, 02, 0d, 00, 14, 02, 02, 0d, 00, 1a, 07, 02, 05, 00, 06]
- file 0 => global file 1
Number of expressions: 2
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 1 operands: lhs = Counter(1), rhs = Expression(0, Sub)
Number of file 0 mappings: 4
- Code(Counter(0)) at (prev + 40, 5) to (start + 1, 24)
- Code(Counter(1)) at (prev + 2, 13) to (start + 0, 20)
- Code(Expression(0, Sub)) at (prev + 2, 13) to (start + 0, 26)
    = (c0 - c1)
- Code(Expression(1, Add)) at (prev + 2, 5) to (start + 0, 6)
    = (c1 + (c0 - c1))

Function name: <try_error_result::Thing2>::call
Raw bytes (28): 0x[01, 01, 02, 01, 05, 05, 02, 04, 01, 33, 05, 01, 18, 05, 02, 0d, 00, 14, 02, 02, 0d, 00, 13, 07, 02, 05, 00, 06]
- file 0 => global file 1
Number of expressions: 2
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 1 operands: lhs = Counter(1), rhs = Expression(0, Sub)
Number of file 0 mappings: 4
- Code(Counter(0)) at (prev + 51, 5) to (start + 1, 24)
- Code(Counter(1)) at (prev + 2, 13) to (start + 0, 20)
- Code(Expression(0, Sub)) at (prev + 2, 13) to (start + 0, 19)
    = (c0 - c1)
- Code(Expression(1, Add)) at (prev + 2, 5) to (start + 0, 6)
    = (c1 + (c0 - c1))
Function name: try_error_result::call
Function name: try_error_result::call
Raw bytes (28): 0x[01, 01, 02, 01, 05, 05, 02, 04, 01, 04, 01, 01, 14, 05, 02, 09, 00, 10, 02, 02, 09, 00, 0f, 07, 02, 01, 00, 02]
- file 0 => global file 1
Number of expressions: 2
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 1 operands: lhs = Counter(1), rhs = Expression(0, Sub)
Number of file 0 mappings: 4
- Code(Counter(0)) at (prev + 4, 1) to (start + 1, 20)
- Code(Counter(1)) at (prev + 2, 9) to (start + 0, 16)
- Code(Expression(0, Sub)) at (prev + 2, 9) to (start + 0, 15)
    = (c0 - c1)
- Code(Expression(1, Add)) at (prev + 2, 1) to (start + 0, 2)
    = (c1 + (c0 - c1))
Function name: try_error_result::main
Function name: try_error_result::main
Raw bytes (28): 0x[01, 01, 02, 01, 05, 05, 02, 04, 01, 70, 01, 02, 0c, 05, 03, 05, 00, 06, 02, 02, 05, 00, 0b, 07, 01, 01, 00, 02]
- file 0 => global file 1
Number of expressions: 2
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 0 operands: lhs = Counter(0), rhs = Counter(1)
- expression 1 operands: lhs = Counter(1), rhs = Expression(0, Sub)
Number of file 0 mappings: 4
- Code(Counter(0)) at (prev + 112, 1) to (start + 2, 12)
- Code(Counter(1)) at (prev + 3, 5) to (start + 0, 6)
- Code(Expression(0, Sub)) at (prev + 2, 5) to (start + 0, 11)
    = (c0 - c1)
- Code(Expression(1, Add)) at (prev + 1, 1) to (start + 0, 2)
    = (c1 + (c0 - c1))
Function name: try_error_result::test1
Function name: try_error_result::test1
Raw bytes (77): 0x[01, 01, 09, 01, 07, 05, 09, 03, 0d, 1d, 11, 16, 1d, 03, 0d, 1f, 0d, 11, 23, 15, 19, 0b, 01, 0c, 01, 02, 17, 03, 07, 09, 00, 0e, 16, 02, 09, 04, 1a, 1d, 06, 0d, 00, 29, 11, 00, 29, 00, 2a, 0e, 01, 0d, 00, 2a, 15, 00, 2a, 00, 2b, 12, 04, 0d, 00, 2a, 19, 00, 2a, 00, 2b, 0d, 03, 05, 00, 0b, 1b, 01, 01, 00, 02]
- file 0 => global file 1
Number of expressions: 9
Number of expressions: 9
- expression 0 operands: lhs = Counter(0), rhs = Expression(1, Add)
- expression 1 operands: lhs = Counter(1), rhs = Counter(2)
- expression 2 operands: lhs = Expression(0, Add), rhs = Counter(3)
- expression 3 operands: lhs = Counter(7), rhs = Counter(4)
- expression 4 operands: lhs = Expression(5, Sub), rhs = Counter(7)
- expression 5 operands: lhs = Expression(0, Add), rhs = Counter(3)
- expression 6 operands: lhs = Expression(7, Add), rhs = Counter(3)
- expression 7 operands: lhs = Counter(4), rhs = Expression(8, Add)
- expression 8 operands: lhs = Counter(5), rhs = Counter(6)
Number of file 0 mappings: 11
- Code(Counter(0)) at (prev + 12, 1) to (start + 2, 23)
- Code(Expression(0, Add)) at (prev + 7, 9) to (start + 0, 14)
    = (c0 + (c1 + c2))
- Code(Expression(5, Sub)) at (prev + 2, 9) to (start + 4, 26)
    = ((c0 + (c1 + c2)) - c3)
- Code(Counter(7)) at (prev + 6, 13) to (start + 0, 41)
- Code(Counter(4)) at (prev + 0, 41) to (start + 0, 42)
- Code(Expression(3, Sub)) at (prev + 1, 13) to (start + 0, 42)
    = (c7 - c4)
- Code(Counter(5)) at (prev + 0, 42) to (start + 0, 43)
- Code(Expression(4, Sub)) at (prev + 4, 13) to (start + 0, 42)
    = (((c0 + (c1 + c2)) - c3) - c7)
- Code(Counter(6)) at (prev + 0, 42) to (start + 0, 43)
- Code(Counter(3)) at (prev + 3, 5) to (start + 0, 11)
- Code(Expression(6, Add)) at (prev + 1, 1) to (start + 0, 2)
    = ((c4 + (c5 + c6)) + c3)
Function name: try_error_result::test2
Function name: try_error_result::test2
Raw bytes (358): 0x[01, 01, 3b, 01, 07, 05, 09, 03, 0d, 41, 11, 4a, 15, 41, 11, 42, 1d, 46, 19, 4a, 15, 41, 11, 4a, 15, 41, 11, 46, 19, 4a, 15, 41, 11, 42, 1d, 46, 19, 4a, 15, 41, 11, 5e, 25, 49, 21, 49, 21, 5e, 25, 49, 21, 8a, 01, 2d, 8e, 01, 29, 92, 01, 41, 03, 0d, 92, 01, 41, 03, 0d, 8e, 01, 29, 92, 01, 41, 03, 0d, 8a, 01, 2d, 8e, 01, 29, 92, 01, 41, 03, 0d, a6, 01, 35, 45, 31, 45, 31, a6, 01, 35, 45, 31, ba, 01, 3d, 4d, 39, 4d, 39, ba, 01, 3d, 4d, 39, c3, 01, 0d, 11, c7, 01, cb, 01, db, 01, 15, cf, 01, d3, 01, d7, 01, 19, 1d, 21, 25, df, 01, e3, 01, 29, 2d, e7, 01, eb, 01, 31, 35, 39, 3d, 28, 01, 3c, 01, 03, 17, 03, 08, 09, 00, 0e, 92, 01, 02, 09, 04, 1a, 41, 06, 0d, 00, 2f, 11, 00, 2f, 00, 30, 4a, 00, 31, 03, 35, 15, 04, 11, 00, 12, 46, 02, 11, 04, 12, 3e, 05, 11, 00, 14, 46, 00, 17, 00, 41, 19, 00, 41, 00, 42, 42, 00, 43, 00, 5f, 1d, 00, 5f, 00, 60, 3e, 01, 0d, 00, 20, 5a, 01, 11, 00, 14, 49, 00, 17, 00, 41, 21, 00, 41, 00, 42, 5e, 00, 43, 00, 60, 25, 00, 60, 00, 61, 5a, 01, 0d, 00, 20, 86, 01, 04, 11, 00, 14, 8e, 01, 00, 17, 00, 42, 29, 00, 42, 00, 43, 8a, 01, 00, 44, 00, 61, 2d, 00, 61, 00, 62, 86, 01, 01, 0d, 00, 20, a2, 01, 01, 11, 00, 14, 45, 00, 17, 01, 36, 31, 01, 36, 00, 37, a6, 01, 01, 12, 00, 2f, 35, 00, 2f, 00, 30, a2, 01, 01, 0d, 00, 20, b6, 01, 01, 11, 00, 14, 4d, 00, 17, 01, 36, 39, 02, 11, 00, 12, ba, 01, 01, 12, 00, 2f, 3d, 01, 11, 00, 12, b6, 01, 02, 0d, 00, 20, 0d, 03, 05, 00, 0b, bf, 01, 01, 01, 00, 02]
- file 0 => global file 1
Number of expressions: 59
Number of expressions: 59
- expression 0 operands: lhs = Counter(0), rhs = Expression(1, Add)
- expression 1 operands: lhs = Counter(1), rhs = Counter(2)
- expression 2 operands: lhs = Expression(0, Add), rhs = Counter(3)
- expression 3 operands: lhs = Counter(16), rhs = Counter(4)
- expression 4 operands: lhs = Expression(18, Sub), rhs = Counter(5)
- expression 5 operands: lhs = Counter(16), rhs = Counter(4)
- expression 6 operands: lhs = Expression(16, Sub), rhs = Counter(7)
- expression 7 operands: lhs = Expression(17, Sub), rhs = Counter(6)
- expression 8 operands: lhs = Expression(18, Sub), rhs = Counter(5)
- expression 9 operands: lhs = Counter(16), rhs = Counter(4)
- expression 10 operands: lhs = Expression(18, Sub), rhs = Counter(5)
- expression 11 operands: lhs = Counter(16), rhs = Counter(4)
- expression 12 operands: lhs = Expression(17, Sub), rhs = Counter(6)
- expression 13 operands: lhs = Expression(18, Sub), rhs = Counter(5)
- expression 14 operands: lhs = Counter(16), rhs = Counter(4)
- expression 15 operands: lhs = Expression(16, Sub), rhs = Counter(7)
- expression 16 operands: lhs = Expression(17, Sub), rhs = Counter(6)
- expression 17 operands: lhs = Expression(18, Sub), rhs = Counter(5)
- expression 18 operands: lhs = Counter(16), rhs = Counter(4)
- expression 19 operands: lhs = Expression(23, Sub), rhs = Counter(9)
- expression 20 operands: lhs = Counter(18), rhs = Counter(8)
- expression 21 operands: lhs = Counter(18), rhs = Counter(8)
- expression 22 operands: lhs = Expression(23, Sub), rhs = Counter(9)
- expression 23 operands: lhs = Counter(18), rhs = Counter(8)
- expression 24 operands: lhs = Expression(34, Sub), rhs = Counter(11)
- expression 25 operands: lhs = Expression(35, Sub), rhs = Counter(10)
- expression 26 operands: lhs = Expression(36, Sub), rhs = Counter(16)
- expression 27 operands: lhs = Expression(0, Add), rhs = Counter(3)
- expression 28 operands: lhs = Expression(36, Sub), rhs = Counter(16)
- expression 29 operands: lhs = Expression(0, Add), rhs = Counter(3)
- expression 30 operands: lhs = Expression(35, Sub), rhs = Counter(10)
- expression 31 operands: lhs = Expression(36, Sub), rhs = Counter(16)
- expression 32 operands: lhs = Expression(0, Add), rhs = Counter(3)
- expression 33 operands: lhs = Expression(34, Sub), rhs = Counter(11)
- expression 34 operands: lhs = Expression(35, Sub), rhs = Counter(10)
- expression 35 operands: lhs = Expression(36, Sub), rhs = Counter(16)
- expression 36 operands: lhs = Expression(0, Add), rhs = Counter(3)
- expression 37 operands: lhs = Expression(41, Sub), rhs = Counter(13)
- expression 38 operands: lhs = Counter(17), rhs = Counter(12)
- expression 39 operands: lhs = Counter(17), rhs = Counter(12)
- expression 40 operands: lhs = Expression(41, Sub), rhs = Counter(13)
- expression 41 operands: lhs = Counter(17), rhs = Counter(12)
- expression 42 operands: lhs = Expression(46, Sub), rhs = Counter(15)
- expression 43 operands: lhs = Counter(19), rhs = Counter(14)
- expression 44 operands: lhs = Counter(19), rhs = Counter(14)
- expression 45 operands: lhs = Expression(46, Sub), rhs = Counter(15)
- expression 46 operands: lhs = Counter(19), rhs = Counter(14)
- expression 47 operands: lhs = Expression(48, Add), rhs = Counter(3)
- expression 48 operands: lhs = Counter(4), rhs = Expression(49, Add)
- expression 49 operands: lhs = Expression(50, Add), rhs = Expression(54, Add)
- expression 50 operands: lhs = Counter(5), rhs = Expression(51, Add)
- expression 51 operands: lhs = Expression(52, Add), rhs = Expression(53, Add)
- expression 52 operands: lhs = Counter(6), rhs = Counter(7)
- expression 53 operands: lhs = Counter(8), rhs = Counter(9)
- expression 54 operands: lhs = Expression(55, Add), rhs = Expression(56, Add)
- expression 55 operands: lhs = Counter(10), rhs = Counter(11)
- expression 56 operands: lhs = Expression(57, Add), rhs = Expression(58, Add)
- expression 57 operands: lhs = Counter(12), rhs = Counter(13)
- expression 58 operands: lhs = Counter(14), rhs = Counter(15)
Number of file 0 mappings: 40
- Code(Counter(0)) at (prev + 60, 1) to (start + 3, 23)
- Code(Expression(0, Add)) at (prev + 8, 9) to (start + 0, 14)
    = (c0 + (c1 + c2))
- Code(Expression(36, Sub)) at (prev + 2, 9) to (start + 4, 26)
    = ((c0 + (c1 + c2)) - c3)
- Code(Counter(16)) at (prev + 6, 13) to (start + 0, 47)
- Code(Counter(4)) at (prev + 0, 47) to (start + 0, 48)
- Code(Expression(18, Sub)) at (prev + 0, 49) to (start + 3, 53)
    = (c16 - c4)
- Code(Counter(5)) at (prev + 4, 17) to (start + 0, 18)
- Code(Expression(17, Sub)) at (prev + 2, 17) to (start + 4, 18)
    = ((c16 - c4) - c5)
- Code(Expression(15, Sub)) at (prev + 5, 17) to (start + 0, 20)
    = ((((c16 - c4) - c5) - c6) - c7)
- Code(Expression(17, Sub)) at (prev + 0, 23) to (start + 0, 65)
    = ((c16 - c4) - c5)
- Code(Counter(6)) at (prev + 0, 65) to (start + 0, 66)
- Code(Expression(16, Sub)) at (prev + 0, 67) to (start + 0, 95)
    = (((c16 - c4) - c5) - c6)
- Code(Counter(7)) at (prev + 0, 95) to (start + 0, 96)
- Code(Expression(15, Sub)) at (prev + 1, 13) to (start + 0, 32)
    = ((((c16 - c4) - c5) - c6) - c7)
- Code(Expression(22, Sub)) at (prev + 1, 17) to (start + 0, 20)
    = ((c18 - c8) - c9)
- Code(Counter(18)) at (prev + 0, 23) to (start + 0, 65)
- Code(Counter(8)) at (prev + 0, 65) to (start + 0, 66)
- Code(Expression(23, Sub)) at (prev + 0, 67) to (start + 0, 96)
    = (c18 - c8)
- Code(Counter(9)) at (prev + 0, 96) to (start + 0, 97)
- Code(Expression(22, Sub)) at (prev + 1, 13) to (start + 0, 32)
    = ((c18 - c8) - c9)
- Code(Expression(33, Sub)) at (prev + 4, 17) to (start + 0, 20)
    = (((((c0 + (c1 + c2)) - c3) - c16) - c10) - c11)
- Code(Expression(35, Sub)) at (prev + 0, 23) to (start + 0, 66)
    = (((c0 + (c1 + c2)) - c3) - c16)
- Code(Counter(10)) at (prev + 0, 66) to (start + 0, 67)
- Code(Expression(34, Sub)) at (prev + 0, 68) to (start + 0, 97)
    = ((((c0 + (c1 + c2)) - c3) - c16) - c10)
- Code(Counter(11)) at (prev + 0, 97) to (start + 0, 98)
- Code(Expression(33, Sub)) at (prev + 1, 13) to (start + 0, 32)
    = (((((c0 + (c1 + c2)) - c3) - c16) - c10) - c11)
- Code(Expression(40, Sub)) at (prev + 1, 17) to (start + 0, 20)
    = ((c17 - c12) - c13)
- Code(Counter(17)) at (prev + 0, 23) to (start + 1, 54)
- Code(Counter(12)) at (prev + 1, 54) to (start + 0, 55)
- Code(Expression(41, Sub)) at (prev + 1, 18) to (start + 0, 47)
    = (c17 - c12)
- Code(Counter(13)) at (prev + 0, 47) to (start + 0, 48)
- Code(Expression(40, Sub)) at (prev + 1, 13) to (start + 0, 32)
    = ((c17 - c12) - c13)
- Code(Expression(45, Sub)) at (prev + 1, 17) to (start + 0, 20)
    = ((c19 - c14) - c15)
- Code(Counter(19)) at (prev + 0, 23) to (start + 1, 54)
- Code(Counter(14)) at (prev + 2, 17) to (start + 0, 18)
- Code(Expression(46, Sub)) at (prev + 1, 18) to (start + 0, 47)
    = (c19 - c14)
- Code(Counter(15)) at (prev + 1, 17) to (start + 0, 18)
- Code(Expression(45, Sub)) at (prev + 2, 13) to (start + 0, 32)
    = ((c19 - c14) - c15)
- Code(Counter(3)) at (prev + 3, 5) to (start + 0, 11)
- Code(Expression(47, Add)) at (prev + 1, 1) to (start + 0, 2)
    = ((c4 + ((c5 + ((c6 + c7) + (c8 + c9))) + ((c10 + c11) + ((c12 + c13) + (c14 + c15))))) + c3)
stderr: none



```

</details>

Zalathar commented:
Rebased and fixed snapshot conflict with #116505. (Hopefully this PR will make coverage mappings less sensitive to inlining.)

Zalathar commented on compiler/rustc_codegen_llvm/src/coverageinfo/map_data.rs:
```diff
@@ -52,16 +49,31 @@ impl<'tcx> FunctionCoverage<'tcx> {
+            if let &CovTerm::Expression(id) = term {
+                expressions_expected_but_not_seen.insert(id);
+            }
+        }
```
If an expression is only used indirectly, we automatically mark it as “seen” when initializing the set. If it ends up being unused, LLVM will delete it for us.

Zalathar replied:
To clarify a bit more:

- If a directly-used expression is detected as having been removed by MIR opts, then all mappings and expression operands referring to that expression will be replaced with zero.
- This can cause indirect-only expressions to become completely unused. We still emit them to LLVM, but LLVM has its own unused-expression pass that deletes them.

Zalathar commented:
More updates from feedback [(diff)](https://github.com/rust-lang/rust/compare/a63ac1bda9e026bd1f89c4a10d9850d9146c2a69..729664c2ab4f550d5ce9bd43e9328fd5e26a589a).

- I reversed the logic of seen expressions so that we just use `expressions_seen`. It turns out that “we treat indirect expressions as seen even though they don't have expression-used statements” is easier to explain than having to juggle double-negatives everywhere.
- Moved the operand-expression assert into `maybe_set_operand_to_zero`.
- Now that we have expression data in function info, I realised that there's no need to ever create a dummy `ffi::CounterExpression`.
- Updated MIR dump syntax *mostly* as requested, though I retained a distinction between expression data (`ExpressionId(..) => Expression { .. }`) and mapping data.

davidtwco commented:
Apologies for not getting to this quicker, I'll re-assign to cjgillot because they've already started reviewing.

r? @cjgillot

bors commented:
:pushpin: Commit 33da0978ac2674d046d3b01a9db33dc7f19339c6 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 33da0978ac2674d046d3b01a9db33dc7f19339c6 -->
<!-- homu: {"type":"Approved","sha":"33da0978ac2674d046d3b01a9db33dc7f19339c6","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 33da0978ac2674d046d3b01a9db33dc7f19339c6 with merge 00d5227707422089acccf0c3071667a9825fced5...
<!-- homu: {"type":"BuildStarted","head_sha":"33da0978ac2674d046d3b01a9db33dc7f19339c6","merge_sha":"00d5227707422089acccf0c3071667a9825fced5"} -->

rust-log-analyzer commented:
The job **`dist-aarch64-apple`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/6564299893/job/17830409792) [(plain)](https://github.com/rust-lang-ci/rust/commit/00d5227707422089acccf0c3071667a9825fced5/checks/17830409792/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6564299893/job/17832578216)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/6564299893/job/17832578216","builder_name":"checks-actions"} -->

cjgillot commented:
@bors retry failed to download from `https://crates.io/api/v1/crates/flate2/1.0.27/download`

bors commented:
:hourglass: Testing commit 33da0978ac2674d046d3b01a9db33dc7f19339c6 with merge cc705b801236d064260bb67b3a0a25e4747fa7ec...
<!-- homu: {"type":"BuildStarted","head_sha":"33da0978ac2674d046d3b01a9db33dc7f19339c6","merge_sha":"cc705b801236d064260bb67b3a0a25e4747fa7ec"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6565257574/job/17836747352)
Approved by: cjgillot
Pushing cc705b801236d064260bb67b3a0a25e4747fa7ec to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"cjgillot","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6565257574/job/17836747352"},"merge_sha":"cc705b801236d064260bb67b3a0a25e4747fa7ec"} -->

rust-timer commented:
Finished benchmarking commit (cc705b801236d064260bb67b3a0a25e4747fa7ec): [comparison URL](https://perf.rust-lang.org/compare.html?start=e1aa9edde0f363fbc6ab4a50c85974306e56667e&end=cc705b801236d064260bb67b3a0a25e4747fa7ec&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=e1aa9edde0f363fbc6ab4a50c85974306e56667e&end=cc705b801236d064260bb67b3a0a25e4747fa7ec&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e1aa9edde0f363fbc6ab4a50c85974306e56667e&end=cc705b801236d064260bb67b3a0a25e4747fa7ec&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.1%  | [3.1%, 3.1%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -4.0% | [-4.0%, -4.0%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=e1aa9edde0f363fbc6ab4a50c85974306e56667e&end=cc705b801236d064260bb67b3a0a25e4747fa7ec&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.8% | [0.8%, 0.8%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.8% | [0.8%, 0.8%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=e1aa9edde0f363fbc6ab4a50c85974306e56667e&end=cc705b801236d064260bb67b3a0a25e4747fa7ec&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1% | [0.0%, 0.2%] | 53    |
| Regressions ❌ <br /> (secondary)  | 0.1% | [0.0%, 0.3%] | 39    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.1% | [0.0%, 0.2%] | 53    |
</details>

**Bootstrap**: 628.593s -> 628.235s (-0.06%)
**Artifact size**: 303.98 MiB -> 304.03 MiB (0.02%)
